pkg/util/patroni/patroni_test.go

```diff
@@ -0,0 +1,71 @@
+package patroni
+
+import (
+	"errors"
+	"fmt"
+	"k8s.io/api/core/v1"
+	"testing"
+)
+
+func newMockPod(ip string) *v1.Pod {
+	return &v1.Pod{
+		Status: v1.PodStatus{
+			PodIP: ip,
+		},
+	}
+}
+
+func TestApiURL(t *testing.T) {
+	var testTable = []struct {
+		podIP            string
+		expectedResponse string
+		expectedError    error
+	}{
+		{
+			"127.0.0.1",
+			fmt.Sprintf("http://127.0.0.1:%d", apiPort),
+			nil,
+		},
+		{
+			"0000:0000:0000:0000:0000:0000:0000:0001",
+			fmt.Sprintf("http://[::1]:%d", apiPort),
+			nil,
+		},
+		{
+			"::1",
+			fmt.Sprintf("http://[::1]:%d", apiPort),
+			nil,
+		},
+		{
+			"",
+			"",
+			errors.New(" is not a valid IP"),
+		},
+		{
+			"foobar",
+			"",
+			errors.New("foobar is not a valid IP"),
+		},
+		{
+			"127.0.1",
+			"",
+			errors.New("127.0.1 is not a valid IP"),
+		},
+		{
+			":::",
+			"",
+			errors.New("::: is not a valid IP"),
+		},
+	}
+	for _, test := range testTable {
+		resp, err := apiURL(newMockPod(test.podIP))
+		if resp != test.expectedResponse {
+			t.Errorf("expected response %v does not match the actual %v", test.expectedResponse, resp)
+		}
+		if err != test.expectedError {
+			if err != nil && test.expectedError != nil && err.Error() != test.expectedError.Error() {
```

Comment:
I'm not sure, why this condition is necessary? If an actual error value is different from expected, then I guess it's a fail even if one of them is `nil`, is that correct?

Reply:
Yeah it's for catching the case where an error was thrown, but none was expected or vice versa.

Reply:
Maybe I'm missing something, but isn't in this case if we expect no errors (`test.expectedError == nil`), but something wrong happened (`err != nil`), then we will still not fail the test, because `(err != nil && test.expectedError != nil) == false`? So that the nested condition is false and `t.Errorf` is not executed?